runsc/boot/gofer_conf.go

```diff
@@ -16,52 +16,174 @@ package boot
+		*u = AnonOverlay
+	default:
+		return fmt.Errorf("invalid gofer mount config upper layer type: %s", v)
+	}
+	return nil
+}
+
+// GoferMountConfLowerType describes how lower layer is configured for the gofer mount.
+type GoferMountConfLowerType byte
+
+const (
+	// NoneLower indicates that this gofer mount has no lower layer.
+	NoneLower GoferMountConfLowerType = iota
+
+	// Lisafs indicates that this gofer mount has a LISAFS lower layer.
+	Lisafs
+
+	// Erofs indicates that this gofer mount has an EROFS lower layer.
+	Erofs
+
+	// LowerMax indicates the number of the valid lower layer types.
+	LowerMax
+)
+
+// String returns a human-readable string representing the lower layer type.
+func (l GoferMountConfLowerType) String() string {
+	switch l {
+	case NoneLower:
+		return "none"
+	case Lisafs:
+		return "lisafs"
+	case Erofs:
+		return erofs.Name
+	}
+	panic(fmt.Sprintf("Invalid gofer mount config lower layer type: %d", l))
+}
+
+// Set sets the value. Set(String()) should be idempotent.
+func (l *GoferMountConfLowerType) Set(v string) error {
+	switch v {
+	case "none":
+		*l = NoneLower
+	case "lisafs":
+		*l = Lisafs
+	case erofs.Name:
+		*l = Erofs
+	default:
+		return fmt.Errorf("invalid gofer mount config lower layer type: %s", v)
+	}
+	return nil
+}
+
+// GoferMountConf describes how a gofer mount is configured in the sandbox.
+type GoferMountConf struct {
+	Upper GoferMountConfUpperType `json:"upper"`
+	Lower GoferMountConfLowerType `json:"lower"`
+}
+
+// String returns a human-readable string representing the gofer mount config.
+func (g GoferMountConf) String() string {
+	return fmt.Sprintf("%s:%s", g.Lower, g.Upper)
+}
+
+// Set sets the value. Set(String()) should be idempotent.
+func (g *GoferMountConf) Set(v string) error {
+	parts := strings.Split(v, ":")
+	if len(parts) != 2 {
+		return fmt.Errorf("invalid gofer mount config format: %q", v)
+	}
+	if err := g.Lower.Set(parts[0]); err != nil {
+		return err
+	}
+	if err := g.Upper.Set(parts[1]); err != nil {
+		return err
+	}
+	if !g.valid() {
+		return fmt.Errorf("invalid gofer mount config: %+v", g)
+	}
+	return nil
+}
+
 // IsFilestorePresent returns true if a filestore file was associated with this.
 func (g GoferMountConf) IsFilestorePresent() bool {
-	return g == SelfOverlay || g == AnonOverlay || g == SelfTmpfs
+	return g.Upper == SelfOverlay || g.Upper == AnonOverlay
 }
 
 // IsSelfBacked returns true if this mount is backed by a filestore in itself.
 func (g GoferMountConf) IsSelfBacked() bool {
-	return g == SelfOverlay || g == SelfTmpfs
+	return g.Upper == SelfOverlay
 }
 
 // ShouldUseOverlayfs returns true if an overlayfs should be applied.
 func (g GoferMountConf) ShouldUseOverlayfs() bool {
-	return g == MemoryOverlay || g == SelfOverlay || g == AnonOverlay
+	return g.Lower != NoneLower && g.Upper != NoOverlay
+}
+
+// ShouldUseTmpfs returns true if a tmpfs should be applied.
+func (g GoferMountConf) ShouldUseTmpfs() bool {
+	return g.Lower == NoneLower
```

Comment:
nit: add comment which says "g.valid() implies that g.Upper != NoOverlay".

Reply:
Will do. That does add clarity. Thanks!